### ISSUE TITLE
[SS 4.0.3 URGENT] DataObject Live version not created

### ISSUE THREAD

**Issue by lucacattide** [tractorcow/silverstripe-fluent#388]
I just noticed a big trouble in my client project that I need to solve **really ASAP**.

Only for DataObjects (*has-many* relation), the DB ```Localised_Live``` table version isn't created during save on CMS-side. This result in ```NULL``` content on the front-end page.
If I log as admin and inspect one of the pages involved (actually same for 3 distinct DataObjects) in draft mode, I see the contents regularly.

Here's a couple of screenshots:

![screen shot 2018-03-29 at 16 01 26 2](https://user-images.githubusercontent.com/12085296/38093435-3fb7554e-336b-11e8-86e1-8f737ac59e4f.png)
![screen shot 2018-03-29 at 16 01 58](https://user-images.githubusercontent.com/12085296/38093638-db151472-336b-11e8-83be-4dec4a3d0c22.png)

Here's the code involved (just picking 1 of 3 - are the same):

DataObject
```
<?php
// Definizione Namespace
use SilverStripe\ORM\DataObject;
use SilverStripe\Forms\FieldList;
use SilverStripe\Assets\Folder;
use SilverStripe\Assets\Image;
use SilverStripe\Forms\RequiredFields;
use SilverStripe\Forms\TextField;
use SilverStripe\AssetAdmin\Forms\UploadField;

/**
 * Classe Headline
 */
class Headline extends DataObject {
    // Dichiarazione Proprietà
    private static $db = [
        'Corpo' => 'HTMLText'
    ];
    private static $has_one = [
        'LogoHeadline' => Image::class,
        'Home' => 'Home'
    ];
    private static $searchable_fields = [
        'Corpo'
    ];
    private static $summary_fields = [
        'Corpo'
    ];
    private static $translate = [
        'Corpo'
    ];

    /**
     * Metogo definizione campi obbligatori
     * Setter
     * @return void
     */
    public function getCMSValidator() {
        return new RequiredFields(array('LogoHeadline'));
    }

    /**
     * Metodo gestione campi CMS
     * Setter
     * @return FieldList $fields Campi Back-End
     */
    public function getCMSfields()
    {
        $cartellaHome = 'home';
        $dimensioneFoto = 0.6 * 1024 * 1024; // 600 Kb
        $uploadFoto = new UploadField('LogoHeadline', 'Carosello (.jpg/.jpeg/.png/.gif - 600 Kb max.)');

        $uploadFoto->getValidator()->setAllowedExtensions(array('jpg', 'jpeg', 'png', 'gif'));
        $uploadFoto->getValidator()->setAllowedMaxFileSize($dimensioneFoto);

        Folder::find_or_make($cartellaHome);

        $uploadFoto->setFolderName($cartellaHome);

        $fields = FieldList::create(
            $uploadFoto,
            TextField::create('Corpo', 'Testo')
        );

        $this->extend('updateCMSFields', $fields);

        return $fields;
    }
}
```

Page:
```
<?php
// Definizione Namespace
use SilverStripe\Forms\FieldList;
use SilverStripe\Assets\Folder;
use SilverStripe\Forms\GridField\GridField;
use SilverStripe\Forms\GridField\GridFieldConfig_RecordEditor;
use SilverStripe\Assets\Image;
use SilverStripe\Forms\TextField;
use SilverStripe\Forms\TextareaField;
use SilverStripe\AssetAdmin\Forms\UploadField;

/**
 * Classe Home
 */
class Home extends Page
{
    // Dichiarazione Proprietà
    private static $db = [
        'Summary' => 'HTMLText',
        'LinkCta' => 'Text'
    ];
    private static $has_many = [
        'Headlines' => 'Headline'
    ];
    private static $many_many = [
        'Cliente' => Image::class
    ];
    private static $translate = [
        'Summary',
        'LinkCta',
        'Headlines'
    ];
    private static $description = 'Home Page';

    /**
     * Metodo gestione campi Back-End
     * Getter
     * @return FieldList $fields Campi Back-End
     */
    public function getCMSFields()
    {
        $fields = Page::getCMSFields();

        $cartellaClienti = 'clienti';
        $dimensioneFoto = 0.6 * 1024 * 1024; // 600 Kb

        $fields->removeFieldFromTab('Root.Main', 'Content');
        $fields->addFieldToTab('Root.Main', TextareaField::create('Summary', 'Riepilogo'));
        $fields->addFieldToTab('Root.Main', TextField::create('LinkCta', 'Link (Call to Action)'));

        $headlineConfig = GridFieldConfig_RecordEditor::create();

        $fields->addFieldToTab('Root.Headline', new GridField('Headlines', 'Headline', $this->Headlines(), $headlineConfig));
        $fields->addFieldToTab('Root.Clienti', $uploadClienti = new UploadField('Cliente', 'Clienti (.jpg/.jpeg/.png/.gif - 600 Kb max.)'));
        $uploadClienti->getValidator()->setAllowedExtensions(array('jpg', 'jpeg', 'png', 'gif'));
        $uploadClienti->getValidator()->setAllowedMaxFileSize($dimensioneFoto);

        Folder::find_or_make($cartellaClienti);

        $uploadClienti->setFolderName($cartellaClienti);

        return $fields;
    }
}
```

YAML
```
---
Name: corporate-consulting
---
SilverStripe\Core\Manifest\ModuleManifest:
  project: corporate-consulting
# Localizzazione
SilverStripe\i18n\i18n:
  default_locale: 'it_IT'
  common_locales:
    it_IT:
      name: 'Italian'
      native: 'Italiano'
    en_GB:
      name: 'English'
      native: 'English'
TractorCow\Fluent\Extension\FluentDirectorExtension:
  detect_locale: false
# Admin/Branding
SilverStripe\Admin\LeftAndMain:
  application_name: 'Corporate Consulting'
  application_link: 'https://corporate.it'
  extra_requirements_css:
    - themes/corporate/css/dist/admin.css
# Immagini
SilverStripe\Core\Injector\Injector:
 SilverStripe\Assets\Image_Backend:
   properties:
     Quality: 75
---
  Name: traduzioni
---
Caso:
  extensions:
    - 'TractorCow\Fluent\Extension\FluentExtension'
Headline:
  extensions:
    - 'TractorCow\Fluent\Extension\FluentExtension'
Reference:
  extensions:
    - 'TractorCow\Fluent\Extension\FluentExtension'
```

I'll be open to any suggestion.

Thanks in advance.

**Comment by dhensby** [comment]
@lucacattide please note it is a 4 day weekend for NZ and the UK so don't expect a prompt response to this.

You can always go to our community resources to look for help: 

There are great [community support options](http://www.silverstripe.org/community/#introduction) where you can get help including [our forum](https://forum.silverstripe.org), [Stack Overflow (remember to tag as silverstripe)](http://stackoverflow.com/questions/tagged/silverstripe) and even [IRC](http://www.silverstripe.org/community/contributing-to-silverstripe/irc-channel/) or [Slack](http://silverstripe.org/slack).

**Comment by lucacattide** [comment]
Thanks @dhensby for the advice, understood. Meanwhile, I found the real cause of the issue. 
Either if the Live tables aren't created in the process, my older SS 3.x habits driven me into this. I'm still studying the full improvement of version 4 (this is the very first project with it). 
During debugging, I noticed that the other pages were suffering the data-missing problem (especially for images). Reading with attention the new fresh SS doc about file management, I learned about the new tricky-version management by declaring the ```$owns``` property [($owns and file versioning)](https://docs.silverstripe.org/en/4/developer_guides/files/file_management/#configuring-file-ownership). In my case:

```
[...]
class Home extends Page
{
    // Dichiarazione Proprietà
    private static $db = [
        'Summary' => 'HTMLText',
        'LinkCta' => 'Text'
    ];
    private static $has_many = [
        'Headlines' => 'Headline'
    ];
    private static $many_many = [
        'Cliente' => Image::class
    ];
    // Here it go
    private static $owns = [
        'Cliente' <---
    ];
    [...]
```

After updating images/pages relations by binding them with it - and, obviously, re-publishing all the contents, all the DataObjects that were considered "drafts" became live versioned. Furthermore, the re-publishing process seems to be necessary at least for DataObjects with associated files (apparently ```$owns``` is a property of ```SiteTree``` only) in order to be considered "published".
I still don't know if the DB table issue is really an "issue" or it always been a part of the module business logic, but for now all seems to works well.
